fedimint-core/src/core.rs

```diff
@@ -120,29 +120,50 @@ macro_rules! module_dyn_newtype_impl_encode_decode {
             ) -> Result<Self, fedimint_core::encoding::DecodeError> {
                 let module_instance_id =
                     fedimint_core::core::ModuleInstanceId::consensus_decode(reader, modules)?;
-                let total_len_u64 = u64::consensus_decode(reader, modules)?;
-                let mut reader = reader.take(total_len_u64);
-                let val = modules
-                    .get(module_instance_id)
-                    .ok_or_else(|| {
-                        fedimint_core::encoding::DecodeError::new_custom(anyhow::anyhow!(
-                            "Module decoder not available: {module_instance_id}"
-                        ))
-                    })?
-                    .decode(&mut reader, module_instance_id, modules)?;
-
-                if reader.limit() != 0 {
-                    return Err(fedimint_core::encoding::DecodeError::new_custom(
-                        anyhow::anyhow!("Dyn type did not consume all bytes during decoding"),
-                    ));
-                }
+                let val = match modules.get(module_instance_id) {
+                    Some(decoder) => {
+                        let total_len_u64 = u64::consensus_decode(reader, modules)?;
+                        let mut reader = reader.take(total_len_u64);
+                        let v = decoder.decode(&mut reader, module_instance_id, modules)?;
+
+                        if reader.limit() != 0 {
+                            return Err(fedimint_core::encoding::DecodeError::new_custom(
+                                anyhow::anyhow!(
+                                    "Dyn type did not consume all bytes during decoding"
+                                ),
+                            ));
+                        }
+
+                        v
+                    }
+                    None => $name::from_typed(
+                        module_instance_id,
+                        $crate::core::DynUnknown(Vec::<u8>::consensus_decode(
+                            reader,
+                            &Default::default(),
+                        )?),
+                    ),
+                };
```

Comment:
This change seems dangerous, we should double-check consensus for reliance on never encountering e.g. consensus items with unknown module instance IDs. Previously these would have been filtered out at the network level, now we'll have to do that later I assume.

Pinging @joschisan

Reply:
Since we rely on `Decodable` on both client and server, I don't think we can avoid this type. Client needs to be able to decode dyn-newtype instances that it doesn't have.

On server this should never happen, but it's a good thing to verify - what attacks a malicious peer could do by injecting unknown dyn-newtypes here.

Notably this type fully validates and preserves information. There can be no ambiguity on encoding level. Every dyn-newtype is `module_instance_id, Vec<u8>`, no matter if it known or not.

Reply:
Sure, I'd just want to avoid suddenly having a way to trigger a panic in our consensus loop.

E.g. this code will panic due to `get_expected`:

https://github.com/fedimint/fedimint/blob/05a03c8f02a31ad5c72c434e5087ac2c0fd98219/fedimint-server/src/consensus/mod.rs#L241-L254

Reply:
@elsirion Fixed.